packages/cell-toolbar/src/celltoolbartracker.ts

```diff
@@ -0,0 +1,372 @@
+
+/**
+ * Class for a cell whose contents overlap with the cell toolbar
+ */
+const TOOLBAR_OVERLAP_CLASS = 'jp-toolbar-overlap';
+
+/**
+ * Icons for use in toolbar.
+ *
+ * These are copied from icon.ts, which is not part of the webpack bundle
+ * because nothing is imported from it.
+ */
+export const addAboveIcon = new LabIcon({
+  name: `${EXTENSION_ID}:add-above`,
+  svgstr: addAboveSvg
+});
+export const addBelowIcon = new LabIcon({
+  name: `${EXTENSION_ID}:add-below`,
+  svgstr: addBelowSvg
+});
+export const deleteIcon = new LabIcon({
+  name: `${EXTENSION_ID}:delete`,
+  svgstr: deleteSvg
+});
+export const duplicateIcon = new LabIcon({
+  name: `${EXTENSION_ID}:duplicate`,
+  svgstr: duplicateSvg
+});
+export const moveDownIcon = new LabIcon({
+  name: `${EXTENSION_ID}:move-down`,
+  svgstr: moveDownSvg
+});
+export const moveUpIcon = new LabIcon({
+  name: `${EXTENSION_ID}:move-up`,
+  svgstr: moveUpSvg
+});
+
+/**
+ * Watch a notebook so that a cell toolbar appears on the active cell
+ */
+export class CellToolbarTracker implements IDisposable {
+  constructor(
+    panel: NotebookPanel,
+    commands: CommandRegistry,
+    settings: ISettingRegistry.ISettings | null
+  ) {
+    this._commands = commands;
+    this._panel = panel;
+    this._settings = settings;
+    this._previousActiveCell = this._panel.content.activeCell;
+
+    if (this._settings) {
+      this._onSettingsChanged();
+      this._settings.changed.connect(this._onSettingsChanged, this);
+    }
+
+    const notebookModel = this._panel.context.model;
+    const cells = notebookModel.cells;
+    cells.changed.connect(this.updateConnectedCells, this);
+
+    // Only add the toolbar to the notebook's active cell (if any) once it has fully rendered and been revealed.
+    panel.revealed.then(() => this._onActiveCellChanged(panel.content));
+
+    // Handle subsequent changes of active cell.
+    panel.content.activeCellChanged.connect(this._onActiveCellChanged, this);
+  }
+
+  _onActiveCellChanged(notebook: Notebook): void {
+    const activeCell = notebook.activeCell;
+    if (!activeCell) {
+      return;
+    }
+
+    if (
+      this._previousActiveCell !== null &&
+      this._previousActiveCell !== undefined
+    ) {
+      this._removeToolbar(this._previousActiveCell.model);
+    }
+    this._addToolbar(activeCell.model);
+    this._previousActiveCell = activeCell;
+
+    this._updateCellForToolbarOverlap(activeCell);
+  }
+
+  get isDisposed(): boolean {
+    return this._isDisposed;
+  }
+
+  dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+    this._isDisposed = true;
+
+    if (this._settings) {
+      this._settings.changed.disconnect(this._onSettingsChanged, this);
+    }
+
+    const cells = this._panel?.context.model.cells;
+    if (cells) {
+      cells.changed.disconnect(this.updateConnectedCells, this);
+      each(cells.iter(), model => this._removeToolbar(model));
+    }
+
+    this._panel = null;
+  }
```

Comment:
Let's call the clear data to be sure.

```suggestion
    Signal.clearData(this);
  }
```